kube-aws-updater

```diff
@@ -246,45 +241,138 @@ cycle_nodes() {
+  # list the nodes in this order: A, B, C, A, B, C.
+  # This way, we avoid draining too many nodes in the same zone sequentially,
+  # which can seriously unbalance the ASG and cause scheduling issues, since we
+  # don't strictly control in which AZ any new nodes will be launched.
+  local old_nodes_by_zone
+  readarray -t old_nodes_by_zone < <(kubectl --context="${kube_context}" get nodes -l "role=${role},retiring=${retire_time}" -ojson |\
+    jq -r '.items | group_by(.metadata.labels["failure-domain.beta.kubernetes.io/zone"])[] | [.[].metadata.name] | join(" ")')
+  local old_nodes
+  declare -t old_nodes
+  local zc="${#old_nodes_by_zone[*]}"
+  local i j
+  i=0
+  for z in "${old_nodes_by_zone[@]}"; do
+    j=0
+    for n in ${z}; do
+      old_nodes[((i + j * zc))]=${n}
+      j=$((j + 1))
+    done
+    i=$((i + 1))
+  done
+  local old_nodes_count="${#old_nodes[*]}"
+
+  # - drain/terminate labelled nodes (sequentially)
+  local nc=$(( asg_count - old_nodes_count + node_batch_size ))
+  for old_node in "${old_nodes[@]}"; do
+    if (( nc < old_nodes_count )) && (( nc % node_batch_size == 0 )); then
+      log "waiting for new nodes to become ready before starting to drain the next node batch"
+      wait_for_ready_nodes "${role}" "${nc}"
+    elif (( nc == old_nodes_count )); then
+      log "final node batch, waiting for final nodes to become ready and disabling the ASG 'Launch' process"
+      wait_for_ready_nodes "${role}" "${nc}"
+      retry aws "${aws_opts[@]}" --output=json autoscaling suspend-processes \
+        --auto-scaling-group-name "${asg_name}" --scaling-processes "Launch"
+    fi
+    kill_node "${old_node}"
+    nc=$((nc + 1))
   done
 }
 
-log "kube cluster: ${kube_context}"
+update() {
+  local role=$1
+  if [[ -z "${resume}" ]]; then
+    label_for_cycling "${role}"
+  fi
 
-if [[ -n "${resume}" ]]; then
-  resume
-  exit 0
-fi
+  local instance_address instance_id asgs asg_name asg_count
+  while : ; do
+    instance_address=$(retry kubectl --context="${kube_context}" get nodes -l "role=${role},retiring=${retire_time}" -o json | jq -r '.items[0].metadata.name')
+    instance_id=$(retry aws "${aws_opts[@]}" --output=json ec2 describe-instances --filters "Name=network-interface.private-dns-name,Values=${instance_address}" |\
+      jq -r '.Reservations[].Instances[].InstanceId')
+    asgs=$(retry aws "${aws_opts[@]}" --output=json autoscaling describe-auto-scaling-groups)
+    asg_name=$(echo "${asgs}" | jq -r ".AutoScalingGroups[] | select(.Instances[].InstanceId==\"${instance_id}\") | .AutoScalingGroupName")
+    asg_count=$(echo "${asgs}" | jq -r ".AutoScalingGroups[] | select(.AutoScalingGroupName==\"${asg_name}\") | .MinSize")
+    if [[ -z "${asg_name}" ]]; then
+        log "could not discover the ASG name, a node might be terminating, sleeping and trying again..."
+        sleep 60
+    else
+        break
+    fi
+  done
+  log "asg_name: '${asg_name}' / asg_count: '${asg_count}'"
```

Comment:
nitpick: this is an odd format.

Although not mentioned, I tried to keep hclog fmt: `message text: key1=value1 key2=value2` or `message text`